WordPress/Classes/Services/PeopleService.swift

```diff
@@ -128,48 +132,52 @@ struct PeopleService {
     ///
     /// - Returns: A new Person instance, with the new Role already assigned.
     ///
-    func updateUser(_ user: User, role: String, failure: ((Error, User) -> Void)?) -> User {
-        var pristineRole: String?
-        var updated: User?
-        coreDataStack.performAndSave { context in
+    func updateUser(_ user: User, role: String, receiveUpdate: @escaping (User) -> Void, failure: ((Error, User) -> Void)?) {
+        coreDataStack.performAndSave({ context throws -> (String, User) in
             guard let managedPerson = managedPersonFromPerson(user, in: context) else {
-                return
+                throw PeopleServiceError.userNotFoundLocally(user)
             }
 
             // OP Reversal
-            pristineRole = managedPerson.role
+            let pristineRole = managedPerson.role
 
             // Pre-emptively update the role
             managedPerson.role = role
 
-            updated = User(managedPerson: managedPerson)
-        }
-
-        // Early exit if failed to update the user
-        guard let updated else {
-            return user
-        }
+            return (pristineRole, User(managedPerson: managedPerson))
+        }, completion: { result in
+            switch result {
+            case let .failure(error):
+                failure?(error, user)
+            case let .success((pristineRole, updated)):
+                receiveUpdate(updated)
+                self.updateRemoteUser(user, role: role, roleToRevertUponFailure: pristineRole, failure: failure)
+            }
+        }, on: .main)
+    }
 
-        // Hit the Backend
+    private func updateRemoteUser(_ user: User, role: String, roleToRevertUponFailure pristineRole: String, failure: ((Error, User) -> Void)?) {
         remote.updateUserRole(siteID, userID: user.ID, newRole: role, success: nil, failure: { error in
-
             DDLogError("### Error while updating person \(user.ID) in blog \(self.siteID): \(error)")
 
-            var reloadedPerson: User!
-            self.coreDataStack.performAndSave { context in
+            self.coreDataStack.performAndSave({ context in
                 guard let managedPerson = self.managedPersonFromPerson(user, in: context) else {
                     DDLogError("### Person with ID \(user.ID) deleted before update")
-                    return
+                    throw PeopleServiceError.userNotFoundLocally(user)
                 }
 
-                managedPerson.role = pristineRole!
+                managedPerson.role = pristineRole
 
-                reloadedPerson = User(managedPerson: managedPerson)
-            }
-            failure?(error, reloadedPerson)
+                return User(managedPerson: managedPerson)
+            }, completion: { result in
+                switch result {
+                case let .failure(error):
+                    failure?(error, user)
```

Comment:
Just to confirm I'm reading the diff correctly, this is a new behavior from the previous code and runs with `error` equal `.userNotFoundLocally` if `managedPersonFromPerson(_:, in:)` doesn't find a match.

In the previous version, instead, the app would have called crashed when `failure?(error, reloadedPerson) because no value would have been assigned to `var reloadedPerson: User!` due to the early return.

Is that correct?

Reply:
The app wouldn't crash, but neither the `success` nor the `failure` block would be called. With this new behaviour, as you noted, the `failures` will be called if couldn't find the `ManagedPerson`.

However, I don't think this error would actually happen though, since this function is called immediately after  the user changes the `ManagerPerson` instance's role on UI, it's not likely that the `ManagedPerson` is deleted in that split second.